### PR TITLE
Allow directives to override ShouldEndSession

### LINKED ISSUE #166: Directives should be able to request alternative ShouldEndSession values
Certain directives require the ShouldEndSession property to be set to a specific value:

* Video Apps require null
* Gadget input requires null
* In Skill Products require true

It would be great if directives could implement an interface that allows them to suggest a ShouldEndSession value.

The getter of the ShouldEndSession property could check for directives implementing this type. If there were none, it would return the manually set value.

If there were registered directives all requiring the same state, then the property returns the requested state (it would also be useful at this stage to have an alternative property - ignored by JSON - that would specify if there were registered directives altering the state)

If multiple directives are registered and they return multiple requested states, then an exception is raised with the details of the directives that are conflicting - allowing the skill builder to alter the response accordingly.

This functionality would allow an easier development cycle for skill builders (either the response is correct for their functionality automatically, or they have an exception explaining why what they're doing can't work) and would also ensure there is enough information to explain what's happening to anyone debugging the skill.

### PR DESCRIPTION
**Pull request by stoiveyp**
This PR deals fixes #166 - allowing directives to override the ShouldEndSession value when it's required. IT does this by introducing IEndSessionDirective as an inherited interface of IDirective (which has the JsonIgnore attribute - this doesn't affect serialization).

The example is the VideoAppDirective - where ShouldEndSession has to be null (otherwise the skill will error on the device).

The code checks if all directives with overrides have the same value - if they do - that value is returned

If there's no override, or the directives conflict about what the override should be, it reverts to the explicit value set against the response (did think about throwing an exception - but realistically in a situation where conflicting values is required - there will probably be a workaround by the team that says what the setting should be)

Other example where this could be used:
    Gadgets need null ShouldEndSession to work properly.
    Purchase flow has to end the session so the device process can take over.

### CODE REVIEW

File: Alexa.NET/Response/Response.cs [hunk 1]
```diff
@@ -14,8 +16,30 @@ public class ResponseBody
         [JsonProperty("reprompt", NullValueHandling = NullValueHandling.Ignore)]
         public Reprompt Reprompt { get; set; }
 
+        private bool? _shouldEndSession = null;
+
         [JsonProperty("shouldEndSession", NullValueHandling = NullValueHandling.Ignore)]
-        public bool? ShouldEndSession { get; set; } = false;
```
Comment (timheuer) [inline Alexa.NET/Response/Response.cs]: This changes effectively now changes a default behavior where ShouldEndSession was previous false as default and now is null as default.  Should we retain the default as false and _shouldEndSession be set to false?

Comment (stoiveyp) [reply Alexa.NET/Response/Response.cs]: Absolutely - existing behavior shouldn't be altered unless a new directive is used.
Updated accordingly :+1:

**Comment by garypretty** [comment]
@timheuer Can this be merged now?  Just checking in as we have some downstream changes to make in our project once your next release is out. Thanks.

**Comment by garypretty** [comment]
Thanks @timheuer @stoiveyp - do you have an idea when the next release will be with this in?

**Comment by stoiveyp** [comment]
Hey @garypretty - changes were released as part of version 1.12.0 :+1:

**Comment by garypretty** [comment]
Ah. Fantastic. Thanks for the heads up!

**Pull request merged**